go/login/login.go

```diff
@@ -0,0 +1,388 @@
+	} else {
+		hostIDType = m.HostIDType
+		hostID = m.HostID
+	}
+
+	if hostIDType != "" {
+		hostIDType += ":"
+	}
+
+	if m.Action != "" {
+		action = "/" + m.Action
+	}
+	return []byte(hostIDType + hostID + action)
+}
+
+// Escapes the string so it can be safely placed inside a URL path segment,
+// replacing "/#?" special characters and not replacing "!*'();:@&=+$,[]" special characters.
+func escape(s string) string {
+	res := url.PathEscape(s)
+	for _, c := range "!*'();:@&=+$,[]" {
+		st := string(c)
+		strings.Replace(res, url.PathEscape(st), st, -1)
+	}
+	return res
+}
+
+// Handshake struct represents the context required for constructing the handshake.
+// It contains Prefix - either service key or its id, UserKey - user's public ephemeral key,
+// MessageTagPrefix - tag calculated under Message.
+type Handshake struct {
+	Prefix           byte
+	UserKey          glome.PublicKey
+	MessageTagPrefix []byte
+}
+
+// URLResponse struct represents the context required for the URL constructing.
+// It contains V - URL format version (currently always 1), HandshakeInfo - handshake info,
+// Msg - message info, d - glome.Dialog for the tag managing.
+type URLResponse struct {
+	V             byte
+	HandshakeInfo Handshake
+	Msg           Message
+	d             *glome.Dialog
+}
+
+// NewResponse returns a new URLResponse corresponding to the given arguments.
+func NewResponse(serviceKeyID uint8, serviceKey glome.PublicKey, userKey glome.PrivateKey,
+	V byte, hostIDType string, hostID string, action string, tagLen uint) (*URLResponse, error) {
+	var prefix byte
+	var r URLResponse
+
+	r.V = V
+
+	d, err := userKey.TruncatedExchange(&serviceKey, 1)
+	if err != nil {
+		return nil, err
+	}
+	r.d = d
+
+	r.Msg = Message{hostIDType, hostID, action}
+
+	if serviceKeyID == 0 {
+		// If no key ID was specified, send the first key byte as the ID.
+		prefix = serviceKey[0] & 0x7f
+	} else {
+		prefix = serviceKeyID & 0x7f
+	}
+	userPublic, err := userKey.Public()
+	if err != nil {
+		return nil, err
+	}
+	HandshakeInfo := Handshake{prefix, *userPublic, r.Tag(tagLen)}
+	r.HandshakeInfo = HandshakeInfo
+
+	return &r, nil
+}
+
+// ValidateAuthCode checks if the received tag corresponding to the base64-url encoded message constructed from the Message.
+// Returns true if the received tag is empty.
+func (r *URLResponse) ValidateAuthCode(tag []byte) bool {
+	if len(tag) == 0 {
+		return true
+	}
+	return r.d.Check(tag, r.Msg.Construct(false), 0)
+}
+
+// Tag returns the tag corresponding to the Msg. The returned tag is calculated with usage of sendingKey.
+func (r *URLResponse) Tag(len uint) []byte {
+	return r.d.Tag(r.Msg.Construct(false), 0)[:len]
+}
+
+// EncToken returns a base64-encoded response token.
+func (r *URLResponse) EncToken() string {
+	return base64.URLEncoding.EncodeToString(r.Tag(glome.MaxTagSize)) // TODO: passing the tag len as param?
+}
+
+// Client side glome-login handler. Should be constructed under NewClient constructor.
```

Comment:
Consider writing prose. "Client implements the client-side of the glome-login protocol." maybe. Given that there is an unexported field it is clear that you must use a constructor and in the public API docs I'd just omit what that is. It should be clear from the documentation. That being said, would it have been possible not to have an unexported field? The constructor does not seem to assert anything. Could we just generate response where we need it, does it make sense to retain it at all given that it likely changes with every message anyway?

Reply:
Updated. What do you mean by "It should be clear from the documentation."? I've mentioned it in the docstring.
About the URLResponse - it is used later for ValidateAuthCode, which should be called on Client rather than on URLResponse (which needs to be done for Server). I guess that is an imperfection of an API and ValidateAuthCode should be also implemented on Server and responses method should become private. Don't know if it worse changing at the moment.